src/TestCase.ts

```diff
@@ -0,0 +1,241 @@
+  command: TestCaseCommand;
+  targets: Target[];
+  languageId: string;
+  marks: DecorationRanges;
+  initialState: TestCaseSnapshot;
+  finalState: TestCaseSnapshot;
+  returnValue: any;
+};
+
+export default class TestCase {
+  talonCommand: string;
+  command: TestCaseCommand;
+  languageId: string;
+  targets: Target[];
+  marks: DecorationRanges;
+  context: TestCaseContext;
+  initialState: TestCaseSnapshot | null = null;
+  finalState: TestCaseSnapshot | null = null;
+  returnValue: any = null;
+
+  constructor(command: TestCaseCommand, context: TestCaseContext) {
+    const activeEditor = vscode.window.activeTextEditor!;
+    const { navigationMap, targets, talonCommand } = context;
+
+    this.talonCommand = talonCommand;
+    this.command = command;
+    this.languageId = activeEditor.document.languageId;
+    this.marks = this.extractTargetedDecorations(targets, navigationMap);
+    this.targets = targets;
+    this.context = context;
+  }
+
+  extractPrimitiveTargetKeys(...targets: PrimitiveTarget[]) {
+    const keys: string[] = [];
+    targets.forEach((target) => {
+      if (target.mark.type === "decoratedSymbol") {
+        const { character, symbolColor } = target.mark;
+        keys.push(NavigationMap.getKey(symbolColor, character));
+      }
+    });
+    return keys;
+  }
+
+  extractTargetKeys(target: Target): string[] {
+    switch (target.type) {
+      case "primitive":
+        return this.extractPrimitiveTargetKeys(target);
+
+      case "list":
+        return target.elements.map(this.extractTargetKeys, this).flat();
+
+      case "range":
+        return this.extractPrimitiveTargetKeys(target.start, target.end);
+
+      default:
+        return [];
+    }
+  }
+
+  extractTargetedDecorations(targets: Target[], navigationMap: NavigationMap) {
+    if (!navigationMap) {
+      return {};
+    }
+
+    const decorationRanges = navigationMap.serializeRanges();
+    const targetedDecorations: DecorationRanges = {};
+    const targetKeys = targets.map(this.extractTargetKeys, this).flat();
+    targetKeys.forEach((key) => {
+      targetedDecorations[key] = decorationRanges[key];
+    });
+    return targetedDecorations;
+  }
+
+  isThatMarkTargeted() {
+    return this.targets.some(
+      (target) => target.type === "primitive" && target.mark.type === "that"
+    );
+  }
+
+  static async getSnapshot(
+    thatMark: SelectionWithEditor[]
+  ): Promise<TestCaseSnapshot> {
+    const activeEditor = vscode.window.activeTextEditor!;
+    return {
+      document: activeEditor.document.getText(),
+      selections: activeEditor.selections.map(serializeSelection),
+      visibleRanges: activeEditor.visibleRanges.map(serializeRange),
+      clipboard: await Clipboard.readText(),
+      thatMark: thatMark.map((mark) => serializeSelection(mark.selection)),
+    };
+  }
+
+  async getSnapshot(): Promise<TestCaseSnapshot> {
+    return await TestCase.getSnapshot(this.context.thatMark.get());
+  }
+
+  async saveSnapshot() {
```

Comment:
The difference between `saveSnapshot`, the two different `getSnapshot` functions, and `writeSnapshot` is a bit confusing.  Might rework / rename